Improve performance of hfe listpack

tezc opened the pull request:
This PR contains a few optimizations for hfe listpack.
- Hfe fields are ordered by TTL in the listpack.  There are two cases that we want to search listpack according to TTLs: 
  - As part of active-expiry, we need to find the fields that are expired. e.g. find fields that have smaller TTLs than given timestamp. 
  - When we want to add a new field, we need to find the correct position to maintain the order by TTL. e.g. find the field that has a higher TTL than the one we want to insert. 
  
  Iterating with lpNext() to compare TTLs has a performance cost as lpNext() calls lpValidateIntegrity() for each entry. Instead, this PR adds `lpFindCb()` to the listpack which accepts a comparator callback. It preserves same validation logic of lpFind() which is faster than search with lpNext(). 
  
- We have field name, value, ttl for a single hfe field. Inserting these items one by one to listpack is costly. Especially, as we place fields according to TTL, most additions will end up in the middle of the listpack. Each insert causes realloc + memmove. This PR introduces `lpBatchInsert()` to add multiple items in one go. 

- For hsetf, if we are going to update value and TTL at the same time, currently, we update the value first and later update the TTL (two distinct listpack operation). This PR improves it by doing it with a single update operation.

sundb commented on src/listpack.c:
```diff
@@ -669,50 +667,35 @@ unsigned char *lpGetValue(unsigned char *p, unsigned int *slen, long long *lval)
-                }
             }
 
+            if (cmp(lp, p, user, value, ll) != 0)
```
if cmp can return zero and non-zero, i prefer to make it return zero as equal, 1 as large, -1 as less.
another alternative is to return 1 as equal, and 0 as non-equal.

tezc replied:
Only concern, if cmp returns large or less, it is useless here as it only cares about equality. I agree zero should mean equal, similar to strcmp().

sundb commented on src/listpack.c:
```diff
@@ -737,21 +720,73 @@ unsigned char *lpFind(unsigned char *lp, unsigned char *p, unsigned char *s,
+                      uint32_t slen, unsigned int skip) {
+    struct lpFindArg arg = {
+            .s = s,
+            .slen = slen
```
```suggestion
        .s = s,
        .slen = slen
```

sundb commented on src/t_hash.c:
```diff
@@ -329,34 +329,59 @@ static void listpackExFree(listpackEx *lpt) {
-        expired++;
-        fptr = lpNext(lpt->lp, fptr);
+        /* Third item of a tuple is expiry time */
+        if (slen == HASH_LP_NO_TTL || (uint64_t) slen >= r->expire_time ) {
```
```suggestion
        if (slen == HASH_LP_NO_TTL || (uint64_t) slen >= r->expire_time) {
```

sundb commented on src/listpack.h:
```diff
@@ -31,6 +33,13 @@ typedef struct {
     /* When integer is used, 'sval' is NULL, and lval holds the value. */
     long long lval;
+
+    /* These variables are used internally by lpInsert() */
+    int enctype;
+    uint64_t enclen;
+    unsigned char intenc[LP_MAX_INT_ENCODING_LEN];
+    unsigned long backlen_size;
+    unsigned char backlen[LP_MAX_BACKLEN_SIZE];
```
i'm sure if we should put these in here instead of adding a new structure like listpackInsertEntry.

tezc replied:
sure, if confusing, we can introduce a new struct

sundb replied:
It looks like `listpackInsertEntry` is wrong, it's used to store the encoded data in the middle of insertion.
perhaps we want another struct, and it's visible for inside.

tezc replied:
I think it is not possible hide it from outside but maybe we can put all these fields into another struct?:

```
typedef struct {
    /* When string is used, it is provided with the length (slen). */
    unsigned char *sval;
    uint32_t slen;
    /* When integer is used, 'sval' is NULL, and lval holds the value. */
    long long lval;

    struct listpackInsertEntry internal; /* Internally used */
} listpackEntry;

sundb replied:
but i didn't see `listpackInsertEntry` was used out listpack.c.
am i missing somethimg?

tezc replied:
`listpackInsertEntry` is not used elsewhere. I meant we can introduce it and add it into `listpackEntry`.  

Maybe I misunderstood you. Do you want to make it totally internal to listpack.c? I think it is not possible without some allocation (either on stack or heap). 

e.g. lpBatchInsert() is called with 10 entries. How do you allocate 10 listpackInsertEntry structs?

sundb replied:
yes, i meant to allocate a dynamic listpackInsertEntry array inside `lpinsertBatch`.

tezc replied:
This is what I was trying to avoid :) I'm not sure if performance cost worth it. Also, it will be regression for the existing code. Maybe we can selectively do it. 

e.g. have stack allocated listpackInsertEntry arr[5] in lpInsertBatch().
If more than 5 entries are given, allocate on heap. 

Still not sure if this is better.

sundb replied:
a static array also works.
i just don't think we should let these encoding data be perceived outside.

sundb commented on src/listpack.c:
```diff
@@ -793,40 +831,54 @@ unsigned char *lpInsert(unsigned char *lp, unsigned char *elestr, unsigned char
     }
 
+
     /* We need to also encode the backward-parsable length of the element
      * and append it to the end: this allows to traverse the listpack from
      * the end to the start. */
```
need to be udpated.

sundb commented on src/listpack.c:
```diff
@@ -860,34 +912,40 @@ unsigned char *lpInsert(unsigned char *lp, unsigned char *elestr, unsigned char
     }
 
-    /* Store the entry. */
+    unsigned char *last = dst; /* Keep pointer to the last added entry */
```
why is the last added entry? if so, we can't continue iterator listapck after indertion or deletion.

tezc replied:
Currently, if you add 'a' to listpack, newp will point to 'a', right?
In the new one, if you add 'a', 'b', it will point 'b'. Isn't that correct? What do you think it should point to? To the first item?

sundb replied:
yes, you're right, maybe we should update the fowlloing comment.
`the element` -> `the last element`?
```
 * If 'newp' is not NULL, at the end of a successful call '*newp' will be set
 * to the address of the element just added, so that it will be possible to
 * continue an interaction with lpNext() and lpPrev().
```

tezc replied:
Ok, sure.

Also, I'm a bit confused as well :) I'll add tests (if missing) for replace, del and add to check newp value

sundb commented on src/listpack.c:
```diff
@@ -961,11 +1019,27 @@ unsigned char *lpAppendInteger(unsigned char *lp, long long lval) {
+ * a single realloc() for all the given entries.
+ *
+ * In each listpackEntry, if 'sval' is  not null, it is assumed entry is string.
+ * Otherwise, 'lval' is used to append integer entry.
+ */
```
```suggestion
 * Otherwise, 'lval' is used to append integer entry. */
```

sundb commented on src/listpack.c:
```diff
@@ -961,11 +1019,27 @@ unsigned char *lpAppendInteger(unsigned char *lp, long long lval) {
+ * In each listpackEntry, if 'sval' is  not null, it is assumed entry is string.
+ * Otherwise, 'lval' is used to append integer entry.
+ */
+unsigned char *lpBatchAppend(unsigned char *lp, listpackEntry *entries, unsigned long len)
+{
```
```suggestion
unsigned char *lpBatchAppend(unsigned char *lp, listpackEntry *entries, unsigned long len) {
```

tezc commented on src/listpack.c:
```diff
@@ -793,40 +831,54 @@ unsigned char *lpInsert(unsigned char *lp, unsigned char *elestr, unsigned char
+        }
+
+        if (where == LP_REPLACE) {
+            assert(rep[0] != LP_EOF);
```
@sundb Another thing, if lp has 1 entry and you call LP_REPLACE with 2 entries, it will trigger the assert. Let me know if you think we should handle it silently. e.g. just delete the existing one and add new ones.

tezc commented:
@sundb I addressed your comments and moved lpBatchInsert() as a separate function. Let's see if we are more comfortable with this one. Maybe we should add a comment that this function might be merged with lpInsert() later in future.

sundb commented:
@tezc yeah, it's clear now.
actually, we can use `lpDeleteRangeWithEntry` instead `lpInsert()` to do the deletion operation, which is more efficient in the future, then we can remove `lpInsert()` at that time.

sundb commented on src/t_hash.c:
```diff
@@ -329,34 +329,59 @@ static void listpackExFree(listpackEx *lpt) {
+/* Callback for lpFindCb(). Used to find number of expired fields as part of
+ * active expiry or when trying to find the position for the new field according
+ * to its expiry time.*/
+static int cbFindInListpack(const unsigned char *lp, unsigned char *p,
+                              void *user, unsigned char *s, long long slen)
```
```suggestion
static int cbFindInListpack(const unsigned char *lp, unsigned char *p,
                            void *user, unsigned char *s, long long slen)
```

sundb commented on src/t_hash.c:
```diff
@@ -329,34 +329,59 @@ static void listpackExFree(listpackEx *lpt) {
+
+    struct lpFingArgs r = {
+            .max_to_search = UINT64_MAX,
+            .expire_time = commandTimeSnapshot(),
```
```suggestion
        .max_to_search = UINT64_MAX,
        .expire_time = commandTimeSnapshot(),
```

sundb commented on src/t_hash.c:
```diff
@@ -386,77 +411,58 @@ static uint64_t listpackExGetMinExpire(robj *o) {
-        serverAssert(!s);
+    struct lpFingArgs r = {
+            .max_to_search = info->maxToExpire,
+            .expire_time = info->now
```
```suggestion
        .max_to_search = info->maxToExpire,
        .expire_time = info->now
```

sundb commented on src/t_hash.c:
```diff
@@ -386,77 +411,58 @@ static uint64_t listpackExGetMinExpire(robj *o) {
+/* Add new field ordered by expire time. */
+void listpackExAddNew(robj *o, sds field, sds value, uint64_t expireAt) {
+    listpackEntry ent[3] = {
+            {.sval = (unsigned char*) field, .slen = sdslen(field)},
+            {.sval = (unsigned char*) value, .slen = sdslen(value)},
+            {.lval = expireAt}
```
```suggestion
        {.sval = (unsigned char*) field, .slen = sdslen(field)},
        {.sval = (unsigned char*) value, .slen = sdslen(value)},
        {.lval = expireAt}
```

sundb reviewed (commented):
LGTM, but not look closely at the t_hash part.

tezc commented on src/listpack.c:
```diff
@@ -745,7 +786,7 @@ unsigned char *lpFind(unsigned char *lp, unsigned char *p, unsigned char *s,
  * by 'p' depending on the 'where' argument, that can be LP_BEFORE, LP_AFTER
  * or LP_REPLACE.
- * 
+ *
```
```suggestion
 * 
```
